File: 338-counting-bits/counting-bits.cpp
class Solution {
public:
    vector<int> countBits(int n) {
       // 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18
       // 0 1 1 2 1 2 
       // k = 4
       if (!n) return {0};
       if (n == 1) return {0, 1};
       vector<int> dp(n + 1, 1);
       int k = 2;

       dp[0] = 0;
       dp[1] = dp[2] = 1;
       for (int i = 3; i <= n; ++i) {
            if (i - k == k)
                k = i;
            else
                dp[i] += dp[i - k];
       }
       return dp;
    }
};

File: 338-counting-bits/counting-bits.rs
impl Solution {
    pub fn count_bits(n: i32) -> Vec<i32> {
        if n == 0 { return vec![0]; }
        if n == 1 { return vec![0, 1]; }
        let mut dp: Vec<i32> = vec![1; n as usize + 1 as usize];
        let mut k: usize = 2;

        dp[0] = 0;
        dp[1] = 1;
        dp[2] = 1;
        for i in 3 ..= n as usize {
            if i - k == k {
                k = i;
            } else {
                dp[i] += dp[i - k];
            }
        }
        
        dp
    }
}